rust/lance-index/src/scalar/inverted/builder.rs

```diff
@@ -0,0 +1,361 @@
+                    &mut self.tokens,
+                    &mut self.invert_list,
+                    &mut self.docs,
+                )
+                .await?;
+            }
+            data_type => {
+                return Err(Error::invalid_input(
+                    format!("unsupported data type {} for inverted index", data_type),
+                    location!(),
+                ))
+            }
+        }
+
+        let token_set_batch = self.tokens.to_batch()?;
+        let mut token_set_writer = dest_store
+            .new_index_file_v2(TOKENS_FILE, token_set_batch.schema())
+            .await?;
+        token_set_writer.write_record_batch(token_set_batch).await?;
+        token_set_writer.finish().await?;
+
+        let invert_list_batch = self.invert_list.to_batch()?;
+        let mut invert_list_writer = dest_store
+            .new_index_file_v2(INVERT_LIST_FILE, invert_list_batch.schema())
+            .await?;
+        invert_list_writer
+            .write_record_batch(invert_list_batch)
+            .await?;
+        invert_list_writer.finish().await?;
+
+        let docs_batch = self.docs.to_batch()?;
+        let mut docs_writer = dest_store
+            .new_index_file_v2(DOCS_FILE, docs_batch.schema())
+            .await?;
+        docs_writer.write_record_batch(docs_batch).await?;
+        docs_writer.finish().await?;
+
+        Ok(())
+    }
+
+    pub async fn remap(
+        &mut self,
+        mapping: &HashMap<u64, Option<u64>>,
+        dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        let tokens_batch = self.tokens.to_batch()?;
+        self.invert_list.remap(mapping);
+        let invert_list_batch = self.invert_list.to_batch()?;
+        self.docs.remap(mapping);
+        let docs_batch = self.docs.to_batch()?;
+
+        let mut token_set_writer = dest_store
+            .new_index_file_v2(TOKENS_FILE, tokens_batch.schema())
+            .await?;
+        token_set_writer.write_record_batch(tokens_batch).await?;
+        token_set_writer.finish().await?;
+
+        let mut invert_list_writer = dest_store
+            .new_index_file_v2(INVERT_LIST_FILE, invert_list_batch.schema())
+            .await?;
+        invert_list_writer
+            .write_record_batch(invert_list_batch)
+            .await?;
+        invert_list_writer.finish().await?;
+
+        let mut docs_writer = dest_store
+            .new_index_file_v2(DOCS_FILE, docs_batch.schema())
+            .await?;
+        docs_writer.write_record_batch(docs_batch).await?;
+        docs_writer.finish().await?;
+
+        Ok(())
+    }
+}
+
+pub async fn update_index<Offset: OffsetSizeTrait>(
+    new_data: SendableRecordBatchStream,
+    token_set: &mut TokenSet,
+    invert_list: &mut InvertedList,
+    docs: &mut DocSet,
+) -> Result<()> {
+    let mut tokenizer = TOKENIZER.clone();
+    let mut stream = new_data;
+    while let Some(batch) = stream.try_next().await? {
+        let doc_col = batch.column(0).as_string::<Offset>();
```

Comment:
IMO it doesn't seem worth making this whole method generic over the offsets. Could we instead write a function that does:

```rust
fn iter_str_array(array: &dyn Array) -> Box<dyn Iterator<Item = Option<&str>> + '_>
```

And iterate over that boxed iterator?